tests/e2e/service_invocation/service_invocation_test.go

```diff
@@ -378,6 +387,188 @@ func TestHeaders(t *testing.T) {
+
+		assert.True(t, ok)
+		assert.NotNil(t, gotSc)
+		assert.Equal(t, expectedTraceID, spanContextToString(gotSc))
 	})
 
+	t.Run("http-to-grpc-tracing", func(t *testing.T) {
+		body, err := json.Marshal(testCommandRequest{
+			RemoteApp:        "grpcapp",
+			Method:           "http-to-grpc-tracing",
+			RemoteAppTracing: "true",
+		})
+		require.NoError(t, err)
+
+		resp, err := utils.HTTPPost(
+			fmt.Sprintf("http://%s/tests/v1_httptogrpctest", externalURL), body)
+		t.Log("checking err...")
+		require.NoError(t, err)
+
+		var appResp appResponse
+		t.Logf("unmarshalling..%s\n", string(resp))
+		err = json.Unmarshal(resp, &appResp)
+
+		var actualHeaders = map[string]string{}
+		json.Unmarshal([]byte(appResp.Message), &actualHeaders)
+		var requestHeaders = map[string][]string{}
+		var responseHeaders = map[string][]string{}
+		json.Unmarshal([]byte(actualHeaders["request"]), &requestHeaders)
+		json.Unmarshal([]byte(actualHeaders["response"]), &responseHeaders)
+
+		require.NoError(t, err)
+
+		assert.NotNil(t, requestHeaders["grpc-trace-bin"][0])
+		assert.Equal(t, 1, len(requestHeaders["grpc-trace-bin"]))
+
+		assert.NotNil(t, responseHeaders["Traceparent"][0])
+		assert.Equal(t, expectedTraceID, responseHeaders["Traceparent"][0])
+	})
+
+	t.Run("grpc-to-http-tracing", func(t *testing.T) {
+		body, err := json.Marshal(testCommandRequest{
+			RemoteApp:        "serviceinvocation-callee-0",
+			Method:           "grpc-to-http-tracing",
+			RemoteAppTracing: "true",
+		})
+		require.NoError(t, err)
+
+		resp, err := utils.HTTPPost(
+			fmt.Sprintf("http://%s/tests/v1_grpctohttptest", externalURL), body)
+		t.Log("checking err...")
+		require.NoError(t, err)
+
+		var appResp appResponse
+		t.Logf("unmarshalling..%s\n", string(resp))
+		err = json.Unmarshal(resp, &appResp)
+
+		var actualHeaders = map[string]string{}
+		json.Unmarshal([]byte(appResp.Message), &actualHeaders)
+		var requestHeaders = map[string][]string{}
+		var responseHeaders = map[string][]string{}
+		json.Unmarshal([]byte(actualHeaders["request"]), &requestHeaders)
+		json.Unmarshal([]byte(actualHeaders["response"]), &responseHeaders)
+
+		require.NoError(t, err)
+
+		assert.NotNil(t, requestHeaders["Traceparent"][0])
+		assert.Equal(t, expectedTraceID, requestHeaders["Daprtest-Traceid"][0])
+
+		assert.NotNil(t, responseHeaders["grpc-trace-bin"][0])
+		assert.Equal(t, 1, len(responseHeaders["grpc-trace-bin"]))
+		traceContext := responseHeaders["grpc-trace-bin"][0]
+
+		t.Logf("received response grpc header..%s\n", traceContext)
+		assert.Equal(t, expectedEncodedTraceID, traceContext)
+		decoded, _ := base64.StdEncoding.DecodeString(traceContext)
+		gotSc, ok := propagation.FromBinary([]byte(decoded))
+
+		assert.True(t, ok)
+		assert.NotNil(t, gotSc)
+		assert.Equal(t, expectedTraceID, spanContextToString(gotSc))
+	})
+}
+
+// spanContextToString returns the SpanContext string representation
+func spanContextToString(sc trace.SpanContext) string {
```

Comment:
this is duplicated.. We can use `SpanContextToW3CString`

Reply:
thanks- earlier having dependency cycle issue, so need to create a helper ..fixed the issue but missed removing this from here and using the diag one..